protocols/noise/src/io.rs

```diff
@@ -215,158 +238,185 @@ impl<T: io::Write> io::Write for NoiseOutput<T> {
-                            return Err(e)
+                        Poll::Ready(Err(e)) => {
+                            return Poll::Ready(Err(e))
+                        }
+                        Poll::Pending => {
+                            this.write_state = WriteState::WriteLen { len, buf, off };
+
+                            return Poll::Pending
                         }
                     }
-                    self.write_state = WriteState::WriteData { len, off: 0 }
+                    this.write_state = WriteState::WriteData { len, off: 0 }
                 }
                 WriteState::WriteData { len, ref mut off } => {
-                    let n = self.io.write(&buffer.write_crypto[*off .. len])?;
+                    let n = match Pin::new(&mut this.io).poll_write(cx, &buffer.write_crypto[*off .. len]) {
+                        Poll::Ready(Ok(n)) => n,
+                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+                        Poll::Pending => return Poll::Pending,
+                    };
                     trace!("flush: wrote {}/{} bytes", *off + n, len);
                     if n == 0 {
                         trace!("flush: eof");
-                        self.write_state = WriteState::Eof;
-                        return Err(io::ErrorKind::WriteZero.into())
+                        this.write_state = WriteState::Eof;
+                        return Poll::Ready(Err(io::ErrorKind::WriteZero.into()))
                     }
                     *off += n;
                     if len == *off {
                         trace!("flush: finished writing {} bytes", len);
-                        self.write_state = WriteState::Init;
-                        return Ok(())
+                        this.write_state = WriteState::Init;
+                        return Poll::Ready(Ok(()))
                     }
                 }
                 WriteState::Eof => {
                     trace!("flush: eof");
-                    return Err(io::ErrorKind::WriteZero.into())
+                    return Poll::Ready(Err(io::ErrorKind::WriteZero.into()))
                 }
-                WriteState::EncErr => return Err(io::ErrorKind::InvalidData.into())
+                WriteState::EncErr => return Poll::Ready(Err(io::ErrorKind::InvalidData.into()))
             }
         }
     }
-}
 
-impl<T: AsyncRead> AsyncRead for NoiseOutput<T> {
-    unsafe fn prepare_uninitialized_buffer(&self, _: &mut [u8]) -> bool {
-        false
-    }
-}
+    fn poll_close(mut self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> futures::task::Poll<std::result::Result<(), std::io::Error>>{
+        Pin::new(&mut self.io).poll_close(cx)
 
-impl<T: AsyncWrite> AsyncWrite for NoiseOutput<T> {
-    fn shutdown(&mut self) -> Poll<(), io::Error> {
-        self.io.shutdown()
     }
+
 }
 
 /// Read 2 bytes as frame length from the given source into the given buffer.
 ///
+// TODO: This is not the case, right?
```

Comment:
> What do you mean by "This is not the case"? Since buf: &mut [u8; 2], passing an offset > 2 will panic since off is used as an index into buf. It does not panic for off == 2 though, I didn't know that. See also rust-lang/rust#32057.

Ah, right, I was a bit confused. Thanks for clarifying!

> That being said, since it's now using AsyncRead, this can of course be done much more conveniently through AsyncReadExt::read_exact, if you're inclined to do that.

@romanb `read_exact` returns a `Future`. Given that `read_frame_len` is called within a non-async function handling this `Future` would be difficult. Please correct me if I am wrong.

Reply:
> > That being said, since it's now using AsyncRead, this can of course be done much more conveniently through AsyncReadExt::read_exact, if you're inclined to do that.
> 
> @romanb `read_exact` returns a `Future`. Given that `read_frame_len` is called within a non-async function handling this `Future` would be difficult. Please correct me if I am wrong.

The uses are inside `AsyncRead` / `AsyncWrite`, so the idea would be to just store the future in `ReadState::ReadLen` (respectively `WriteState::WriteLen`) instead of `buf` + `off` and polling it from within `poll_read` / `poll_write`. Where do you see the difficulties?